README.md
# DockDuck

![img/duck-logo.png](img/duck-logo.png)

[简体中文](README_zh-CN.md)

DockDuck adds a development kit layer to a specific docker image, such as starting openssh-server and installing gdbserver, and provides some scripts as shortcuts for using docker.

It has the following features:

1) **Support adding development kits to docker images to generate dock-duck images**.
2) **Support logging into dock-duck container through SSH protocol**.
3) **Support logging into dock-duck containers as the same user as the host**.
4) **Provide dock-duck tools to use dock-duck images quickly**.

## 1. Prerequisites

- [GNU Bash](https://www.gnu.org/software/bash/)
- [UNIX-like OS](https://en.wikipedia.org/wiki/Unix-like)

## 2. Quick Start

Download suitable [releases](https://github.com/duruyao/DockDuck/releases) for your OS (currently only supports UNIX-like OS).

### 2.1. Build New DockDuck Image

There are some pre-built dock-duck images in the directory [duruyao/vimicro-ai](https://hub.docker.com/r/duruyao/vimicro-ai).

Modify the [Dockerfile](./Dockerfile) according to your needs before building.

```shell
bash build.sh [--name NEW_NAME] IMAGE
```

### 2.2. Install DockDuck Tools

```shell
bash install.sh [DIRECTORT]
```

Reload Bash configuration.

```bash
source ~/.bashrc
```

### 2.3. Use DockDuck Tools

Type `dk -h` or `dk --help` to show usage:

```shell
Usage: dk [OPTIONS] [COMMAND]

Tools for using dock-duck containers

Options:
  -h, --help                    Display this help message
  -v, --version                 Print version information and quit

Commands:
  attach                        Attach to a running dock-duck container
  detach                        Launch a new dock-duck container in the background
  images                        List images
  ps                            List running dock-duck containers
  run                           Run a command in a new dock-duck container
  stop                          Stop one or more running containers
  uninstall                     Uninstall dock-duck tools
  update                        Update dock-duck tools

Run 'dk COMMAND --help' for more information on a command

See more about DockDuck at https://github.com/duruyao/DockDuck

```

### 2.4. Update DockDuck Tools

```shell
dk update
```

### 2.5. Uninstall DockDuck Tools

```shell
dk uninstall
```


CHANGELOG.md
# [1.1.0](https://github.com/duruyao/DockDuck/compare/v1.0.0...v1.1.0) (2022-05-16)


### Bug Fixes

* change ports mapping ([de93c66](https://github.com/duruyao/DockDuck/commit/de93c6676315e0a81b27b0a8538f5ac92e7a11b9)), closes [#6](https://github.com/duruyao/DockDuck/issues/6)
* fix bug that writing other containers to data file ([7410f05](https://github.com/duruyao/DockDuck/commit/7410f0574ca624c1024477f391722e015b026b9a))
* fix method of initializing seed ([72fc995](https://github.com/duruyao/DockDuck/commit/72fc995d1d01c09f6c1a59105d02549f0003b909)), closes [#7](https://github.com/duruyao/DockDuck/issues/7)
* update the tools `dk-attach` and `dk-stop` ([567c1b5](https://github.com/duruyao/DockDuck/commit/567c1b5b7d54576c61cb0845469a1ac8e76f0301)), closes [#5](https://github.com/duruyao/DockDuck/issues/5)
* write a non-existed container name to the empty data file ([ba54a92](https://github.com/duruyao/DockDuck/commit/ba54a925f27402ea2ff6360396742bbf20c12cee)), closes [#8](https://github.com/duruyao/DockDuck/issues/8)


### Features

* add debug mode for developer ([e6f33eb](https://github.com/duruyao/DockDuck/commit/e6f33eb5fe06adf67107298a65bc0f0200b17ecf))
* automatically generate unique container name ([01560e9](https://github.com/duruyao/DockDuck/commit/01560e9c5ae1171b63742824330e69843061bb8d))


### Performance Improvements

* search images and containers by using `docker <ps|images> --format STRING --filter FILTER` ([6237ba1](https://github.com/duruyao/DockDuck/commit/6237ba1a7fb69df194bd3fa92c1499da32ec36bd))



# [1.0.0](https://github.com/duruyao/DockDuck/compare/v0.9.0...v1.0.0) (2022-05-13)


### Bug Fixes

* handle non-word-breaking characters (`:`) in completion words ([b70b129](https://github.com/duruyao/DockDuck/commit/b70b129a7c0f4823f39e489ab84358b864bc19a3)), closes [#3](https://github.com/duruyao/DockDuck/issues/3)



# 0.9.0 (2022-05-13)


### Bug Fixes

* add console print functions for all scripts ([c15764b](https://github.com/duruyao/DockDuck/commit/c15764b0690109443b3e12d153a0d9087df3f521)), closes [#4](https://github.com/duruyao/DockDuck/issues/4)
* basic image name contains "/" ([32c11c2](https://github.com/duruyao/DockDuck/commit/32c11c2747e091dfb66dcfa64d3f6bfdda097f64))
* bug(s) extracting repository and tag from docker image ([0db2e68](https://github.com/duruyao/DockDuck/commit/0db2e6814c288aa165b37c4c6a71cb310e3d1ed3))
* cannot insert a line to the 1st line of a empty file ([8e60175](https://github.com/duruyao/DockDuck/commit/8e60175e396bd40fe529e1b69ce35d4404bdb0b0))
* change `dk-complete` path in the file `~/.bashrc` ([7be6202](https://github.com/duruyao/DockDuck/commit/7be62020d81dd7f70438ad78aa56f2de1497ceda))
* check image exists by using command `docker images -q IMAGE` ([778cb1e](https://github.com/duruyao/DockDuck/commit/778cb1e5d3e911928804d67994c470d88c3581d3)), closes [#1](https://github.com/duruyao/DockDuck/issues/1)
* default NEW_IMAGE cover passed NEW_IMAGE ([e4c7145](https://github.com/duruyao/DockDuck/commit/e4c714567cc32e7410b3ced566883a9127d502a5))
* env vars missing for normal users ([31fb6e9](https://github.com/duruyao/DockDuck/commit/31fb6e9cf65de2457eac4870ea149398c4d0488a))
* fix bug that decimal plus hex ([f325735](https://github.com/duruyao/DockDuck/commit/f325735bc4fb7853f2d8602ec89d05dac635b86e))
* fix bug: DK_HOME: unbound variable ([b9b287f](https://github.com/duruyao/DockDuck/commit/b9b287fe57e26df12ad2097756eaf2e694c63874))
* fix error passing args ([4405fba](https://github.com/duruyao/DockDuck/commit/4405fbaf0b39001bd24ef54bc3e4511cf6b626af))
* fix string 'DockDucK' to 'DockDuck' ([85bc86c](https://github.com/duruyao/DockDuck/commit/85bc86c7d2da79a3b6be7d7a7b03207e006c5b9c))
* refresh path/to/containers.txt before `dk ps` ([34884c9](https://github.com/duruyao/DockDuck/commit/34884c94d4d476edf5b5d5a9d4b6510172ffc9dd)), closes [#2](https://github.com/duruyao/DockDuck/issues/2)
* unable to terminate while loop of parsing args ([2b1860c](https://github.com/duruyao/DockDuck/commit/2b1860cdcd2f7f1ed87b1b810d7e5687d6678a35))


### Features

* add 000-todo-all.sh ([45600af](https://github.com/duruyao/DockDuck/commit/45600af77531eb8866bdb0281cefd7418ddd3548))
* add build.sh for building docker image according an existed image ([d7864b7](https://github.com/duruyao/DockDuck/commit/d7864b7966110228ac0b1bd44ccd4f30fee29835))
* add cli to update dk tools ([3463929](https://github.com/duruyao/DockDuck/commit/3463929e7913105bf2b3354f53391d0744512f0f))
* add dk-run (unfinished) ([a3c14af](https://github.com/duruyao/DockDuck/commit/a3c14af2b697e411137c2a67585a0d3eadbeb560))
* add install.sh for installing tools of DockerDK ([2fbbf8f](https://github.com/duruyao/DockDuck/commit/2fbbf8fe232598c33f3195c8f959155531e79363))
* add more commands in Dockerfile ([f663c3c](https://github.com/duruyao/DockDuck/commit/f663c3c04c148804b438347a22ae4062975adba1))
* add sub command 'images' ([11fcb89](https://github.com/duruyao/DockDuck/commit/11fcb8913914e3874edd44adb858438f5a8e8ee3))
* add uninstall tool ([325f6d3](https://github.com/duruyao/DockDuck/commit/325f6d3a0096f0080b4d91c157bfeee0e9081dc7))
* change version id to v2022.03.18 ([ccec2a4](https://github.com/duruyao/DockDuck/commit/ccec2a41eae1f0596a1491e550c1d34d8926a342))
* check if docker exist ([776c40f](https://github.com/duruyao/DockDuck/commit/776c40f58b3459740938bfbe221d824bc849c0c6))
* check sth and show warning info ([6092c14](https://github.com/duruyao/DockDuck/commit/6092c1476367881df4b29173b1cd52d5d531a032))
* copy tools/* to DK_HOME/ ([505e7d5](https://github.com/duruyao/DockDuck/commit/505e7d5bc70cede33c5fd421b346785f20f6b863))
* deprecate the risky command option '--other-flags' ([f592610](https://github.com/duruyao/DockDuck/commit/f592610c5ade6d0155120608fcde6280320be452))
* dk as the entry of application ([682ecf1](https://github.com/duruyao/DockDuck/commit/682ecf18a5c28ce6c6f21376fb0180c970978c41))

build.sh
#!/usr/bin/env bash

##  Copyright (c) 2022, DURUYAO. All rights reserved.
##
##  Licensed under the Apache License, Version 2.0 (the "License");
##  you may not use this file except in compliance with the License.
##  You may obtain a copy of the License at
##
##      http://www.apache.org/licenses/LICENSE-2.0
##
##  Unless required by applicable law or agreed to in writing, software
##  distributed under the License is distributed on an "AS IS" BASIS,
##  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
##  See the License for the specific language governing permissions and
##  limitations under the License.

## date:   2022-01-20
## author: duruao@gmail.com
## desc:   build a new dock-duck image with one more development kit layer than the base image

set -euo pipefail

function error_ln() {
  printf "\033[1;32;31m%s\n\033[m" "$1"
}

function warning_ln() {
  printf "\033[1;33m%s\n\033[m" "$1"
}

function info_ln() {
  printf "\033[0;32;32m%s\n\033[m" "$1"
}

function show_usage() {
  cat <<EOF
Usage: $0 [OPTIONS] IMAGE

Build a new dock-duck image with one more development kit layer than the base image

Options:
  -h, --help                    Display this help message
      --name STRING             Assign a name to the new image (default: "IMAGE:TAG-dk")

See more about DockDuck at https://github.com/duruyao/DockDuck

EOF
}

image=""
name=""

if [ -z "$(command -v docker)" ]; then
  error_ln "Error: 'docker' required but not found" >&2
  show_usage >&2
  exit 1
fi

## parse arguments
while (($#)); do
  case "$1" in
  -h | --help)
    show_usage
    exit 0
    ;;

  --name)
    name="$2"
    shift 2
    ;;

  --* | -*)
    error_ln "Error: Unknown flag: '$1'" >&2
    show_usage >&2
    exit 1
    ;;

  *)
    repo="${1//\:*/}"
    tag="${1//*\:/}"
    if ! echo "$1" | grep -q ":"; then
      tag="latest"
    fi
    if [ -z "$(docker images -q "${repo}:${tag}")" ]; then
      error_ln "Error: No such docker image: '${repo}:${tag}'" >&2
      show_usage >&2
      exit 1
    fi
    image="$1"
    if [ -z "${name}" ]; then
      name="${repo}:${tag}-dk"
    fi
    shift 1
    ;;
  esac
done

if [ -z "${image}" ]; then
  error_ln "Error: Missing IMAGE" >&2
  show_usage >&2
  exit 1
fi

## change Dockerfile
sed -i "/FROM .*/d" "${PWD}"/Dockerfile
sed -i "1i FROM ${image}" "${PWD}"/Dockerfile

set -x
docker build -t "${name}" -f "${PWD}"/Dockerfile "${PWD}"


uninstall.sh
#!/usr/bin/env bash

##  Copyright (c) 2022, DURUYAO. All rights reserved.
##
##  Licensed under the Apache License, Version 2.0 (the "License");
##  you may not use this file except in compliance with the License.
##  You may obtain a copy of the License at
##
##      http://www.apache.org/licenses/LICENSE-2.0
##
##  Unless required by applicable law or agreed to in writing, software
##  distributed under the License is distributed on an "AS IS" BASIS,
##  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
##  See the License for the specific language governing permissions and
##  limitations under the License.

## date:   2022-01-20
## author: duruao@gmail.com
## desc:   uninstall dock-duck from $DK_HOME

set -euo pipefail

function error_ln() {
  printf "\033[1;32;31m%s\n\033[m" "$1"
}

function warning_ln() {
  printf "\033[1;33m%s\n\033[m" "$1"
}

function info_ln() {
  printf "\033[0;32;32m%s\n\033[m" "$1"
}

function show_usage() {
  cat <<EOF
Usage: $0 [OPTIONS]

Uninstall dock-duck from \$DK_HOME

Options:
  -h, --help                    Display this help message

See more about DockDuck at https://github.com/duruyao/DockDuck

EOF
}

## parse arguments
while (($#)); do
  case "$1" in
  -h | --help)
    show_usage
    exit 0
    ;;

  --* | -* | *)
    error_ln "Error: Unknown flag: '$1'" >&2
    show_usage >&2
    exit 1
    ;;
  esac
done

echo "Uninstalling ..."
echo

if [ "NOTFOUND" == "${DK_HOME-NOTFOUND}" ]; then
  warning_ln "Warning: \$DK_HOME not found"
  exit
fi

# shellcheck disable=SC2153
dk_home="${DK_HOME}"
dk_data_home="${dk_home}/data"
dk_tools_home="${dk_home}/tools"

rm -f "${dk_home}"/dk
rm -rf "${dk_data_home}"/*.logo
rm -rf "${dk_tools_home}"

line="export DK_HOME=.*"
sed -i "/${line}/d" "${HOME}"/.bashrc

line="export PATH=\"\$DK_HOME:\$PATH\""
sed -i "/${line}/d" "${HOME}"/.bashrc

line="source \$DK_HOME.*"
sed -i "/${line}/d" "${HOME}"/.bashrc

source "${HOME}"/.bashrc

echo "Uninstall dock-duck from ${dk_home}"


dk-install.sh
#!/usr/bin/env bash

##  Copyright (c) 2022, DURUYAO. All rights reserved.
##
##  Licensed under the Apache License, Version 2.0 (the "License");
##  you may not use this file except in compliance with the License.
##  You may obtain a copy of the License at
##
##      http://www.apache.org/licenses/LICENSE-2.0
##
##  Unless required by applicable law or agreed to in writing, software
##  distributed under the License is distributed on an "AS IS" BASIS,
##  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
##  See the License for the specific language governing permissions and
##  limitations under the License.

## date:   2022-01-20
## author: duruao@gmail.com
## desc:   install dock-duck to the specified directory

set -euo pipefail

function error_ln() {
  printf "\033[1;32;31m%s\n\033[m" "$1"
}

function warning_ln() {
  printf "\033[1;33m%s\n\033[m" "$1"
}

function info_ln() {
  printf "\033[0;32;32m%s\n\033[m" "$1"
}

function show_usage() {
  cat <<EOF
Usage: $0 [OPTIONS]

Install dock-duck to the specified directory

Options:
  -h, --help                    Display this help message
      --prefix STRING           Specify an installation directory (default: "${HOME}/.duck")

See more about DockDuck at https://github.com/duruyao/DockDuck

EOF
}

prefix="${HOME}/.duck"

## parse arguments
while (($#)); do
  case "$1" in
  -h | --help)
    show_usage
    exit 0
    ;;

  --prefix)
    if [ -z "$2" ]; then
      error_ln "Error: '$1' requires a non empty argument" >&2
      show_usage >&2
      exit 1
    elif [ ! -d "$2" ]; then
      error_ln "Error: No such directory: '$2'" >&2
      show_usage >&2
      exit 1
    fi
    prefix="$2"
    shift 2
    ;;

  --* | -* | *)
    error_ln "Error: Unknown flag: '$1'" >&2
    show_usage >&2
    exit 1
    ;;
  esac
done

chmod +x "${PWD}"/app/*
chmod +x "${PWD}"/tools/*

dk_home="${prefix}"
dk_data_home="${dk_home}/data"
dk_tools_home="${dk_home}/tools"

if [ -f "${dk_home}/dk" ] || [ -d "${dk_tools_home}" ]; then
  warning_ln "Warning: There may already be a dock-duck ($("${dk_home}"/dk -v | sed "s/.* //")) in ${dk_home}"
  # shellcheck disable=SC2162
  read -p "Do you want to continue installing the current version? [Y/n] " install_dk
  # shellcheck disable=SC2143
  if [ -n "$(echo "${install_dk}" | grep -E "n|no|N|No|NO")" ]; then
    exit
  fi
  echo ""
fi

echo "Installing ..."
echo

mkdir -p "${dk_home}"
mkdir -p "${dk_data_home}"
mkdir -p "${dk_tools_home}"

cp "${PWD}"/app/* "${dk_home}"/
cp "${PWD}"/tools/* "${dk_tools_home}"/
cp "${PWD}"/img/*.logo "${dk_data_home}"/

line="export DK_HOME=\"${dk_home}\""
if ! grep -Fxq "${line}" "${HOME}"/.bashrc; then
  echo "${line}" >>"${HOME}"/.bashrc
fi

line="export PATH=\"\$DK_HOME:\$PATH\""
if ! grep -Fxq "${line}" "${HOME}"/.bashrc; then
  echo "${line}" >>"${HOME}"/.bashrc
fi

line="source \$DK_HOME/tools/dk-complete" # FIXME: use variable instead of value
if ! grep -Fxq "${line}" "${HOME}"/.bashrc; then
  echo "${line}" >>"${HOME}"/.bashrc
fi

source "${HOME}"/.bashrc

echo "Install dock-duck to ${dk_home}"
if [ -n "$(command -v tree)" ]; then
  tree "${dk_home}"
else
  ls "${dk_home}"
fi

echo ""
echo "Try 'source ~/.bashrc' to reload PATH"

install.sh
#!/usr/bin/env bash

##  Copyright (c) 2022, DURUYAO. All rights reserved.
##
##  Licensed under the Apache License, Version 2.0 (the "License");
##  you may not use this file except in compliance with the License.
##  You may obtain a copy of the License at
##
##      http://www.apache.org/licenses/LICENSE-2.0
##
##  Unless required by applicable law or agreed to in writing, software
##  distributed under the License is distributed on an "AS IS" BASIS,
##  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
##  See the License for the specific language governing permissions and
##  limitations under the License.

## date:   2022-01-20
## author: duruao@gmail.com
## desc:   install dock-duck to the specified directory

set -euo pipefail

function error_ln() {
  printf "\033[1;32;31m%s\n\033[m" "$1"
}

function warning_ln() {
  printf "\033[1;33m%s\n\033[m" "$1"
}

function info_ln() {
  printf "\033[0;32;32m%s\n\033[m" "$1"
}

function show_usage() {
  cat <<EOF
Usage: $0 [OPTIONS]

Install dock-duck to the specified directory

Options:
  -d, --debug                   Install a debug version
  -h, --help                    Display this help message
      --prefix STRING           Specify an installation directory (default: "${HOME}/.duck")

See more about DockDuck at https://github.com/duruyao/DockDuck

EOF
}

prefix="${HOME}/.duck"
debug=false

## parse arguments
while (($#)); do
  case "$1" in
  -d | --debug)
    debug=true
    shift 1
    ;;

  -h | --help)
    show_usage
    exit 0
    ;;

  --prefix)
    if [ -z "$2" ]; then
      error_ln "Error: '$1' requires a non empty argument" >&2
      show_usage >&2
      exit 1
    elif ! [ -d "$2" ]; then
      error_ln "Error: No such directory: '$2'" >&2
      show_usage >&2
      exit 1
    fi
    prefix="$2"
    shift 2
    ;;

  --* | -* | *)
    error_ln "Error: Unknown flag: '$1'" >&2
    show_usage >&2
    exit 1
    ;;
  esac
done

chmod +x "${PWD}"/app/*
chmod +x "${PWD}"/tools/*

dk_home="${prefix}"
dk_app_home="${dk_home}/app"
dk_img_home="${dk_home}/img"
dk_data_home="${dk_home}/data"
dk_tools_home="${dk_home}/tools"

if [ -d "${dk_app_home}" ] || [ -d "${dk_img_home}" ] || [ -d "${dk_tools_home}" ]; then
  warning_ln "Warning: There may already be a old dock-duck in ${dk_home}"
  # shellcheck disable=SC2162
  read -p "Do you want to continue installing the current version? [Y/n] " install_dk
  # shellcheck disable=SC2143
  if [ -n "$(echo "${install_dk}" | grep -E "n|no|N|No|NO")" ]; then
    exit
  fi
  echo ""
fi

echo "Installing ..."
echo

mkdir -p "${dk_home}"
mkdir -p "${dk_app_home}"
mkdir -p "${dk_img_home}"
mkdir -p "${dk_data_home}"
mkdir -p "${dk_tools_home}"

cp "${PWD}"/app/* "${dk_app_home}"
cp "${PWD}"/img/* "${dk_img_home}"
cp "${PWD}"/tools/* "${dk_tools_home}"

if ${debug}; then
  rm -f "${dk_app_home}"/*
  ln -s "${PWD}"/app/dk "${dk_app_home}"/dk
  rm -f "${dk_tools_home}"/*
  # shellcheck disable=SC2207
  dk_tools=($(ls ./tools))
  for dk_tool in "${dk_tools[@]}"; do
    ln -s "${PWD}"/tools/"${dk_tool}" "${dk_tools_home}"/"${dk_tool}"
  done
fi

bash_config="# >>> DockDuck Bash Config Begin >>>
export DK_HOME=\"${dk_home}\" # DockDuck Bash Config
export PATH=\"\$DK_HOME/app:\$PATH\" # DockDuck Bash Config
source \"\$DK_HOME/tools/dk-complete\" # DockDuck Bash Config
# <<< DockDuck Bash Config End   <<<"

sed -i "/DockDuck Bash Config/d" "${HOME}"/.bashrc
echo "${bash_config}" >>"${HOME}"/.bashrc

echo "Install dock-duck to ${dk_home}"
if [ -n "$(command -v tree)" ]; then
  tree "${dk_home}"
else
  ls "${dk_home}"
fi

echo ""
echo "Try 'source ~/.bashrc' to reload PATH"
